scripts/weekly-summary

```diff
@@ -75,20 +76,49 @@ def get_artifacts(token, artifact_names, start, end):
       "expires_at": "2022-03-25T03:29:06Z"
     },
     """
-    req = Request(URL)
-    req.add_header("Accept", "application/vnd.github.v3+json")
-    req.add_header("Authorization", f"token {token}")
-    with urlopen(req) as r:
-        data = json.load(r)
 
-    # Filter out the artifacts within the date range and names
     artifacts = []
-    for a in data["artifacts"]:
-        if a["name"] not in artifact_names:
-            continue
-        updated_at = datetime.fromisoformat(a["updated_at"][:-1])
-        if start <= updated_at <= end:
-            artifacts.append(a)
+    page = 1
+    retry_limit = 3
+    while True:
+        req = Request(URL + f"&page={page}")
+        req.add_header("Accept", "application/vnd.github.v3+json")
+        req.add_header("Authorization", f"token {token}")
+        with urlopen(req) as r:
+            # Handle hitting the GitHub rate limit
+            # If the reset time is < 90s in the future, wait for it (trying 3 times)
+            # Otherwise raise an error
+            if r.status == 403:
+                try:
+                    reset = int(r.headers.get("X-RateLimit-Reset"))
+                except:
+                    raise RuntimeError("Hit GitHub rate limit. Reset header missing.")
+                if retry_limit == 0 or time.time() > reset or reset - time.time() > 90:
+                    raise RuntimeError("Hit GitHub rate limit. Reset is at %s" % time.ctime(reset))
+
+                # Try waiting until the reset time
+                time.sleep(reset - time.time())
+                retry_limit = retry_limit - 1
+                continue
+
+            if r.status != 200:
+                raise RuntimeError("Error (%d) with API request: %s" % (r.status, str(r)))
+
+            data = json.load(r)
+
+        # Only include the artifacts within the date range and names
+        for a in data["artifacts"]:
+            if a["name"] not in artifact_names:
+                continue
+            updated_at = datetime.fromisoformat(a["updated_at"][:-1])
+            if start <= updated_at <= end:
+                artifacts.append(a)
+
+        if len(data["artifacts"]) < 100:
+            break
```

Comment:
Is this the right exit condition? Or rather, is this actually a sensible way to query the API? AIUI, the way it's written here, this tries to get all pages until last which is detected by having less than 100 results, or we keep going. So I see the following failure modes:
- If there are 32768 pages, we get 32768. I tried the query and we now have 37 pages that return something, starting at `2020-11-12`, whatever that means. Is there a way to limit this better? Surely not all 37 pages are needed to get the last week's data?
- If the last page is perfectly aligned by coincidence, which *will* happen one day, this breaks when trying to get the next nonexistent page. We get status 304 on things beyond end, which as coded dies due to `304 != 200`.

Reply:
Uh, it's not guaranteed to be ordered? That's nasty.

At any rate, hitting the limits very much depends if downloading the archives counts as API request or not.

The limit for authenticated api queries seems to be 5000 (couldn't find any exact number), and we'd be making the requests with the token, so we probably need to fit in that limit. 37 pages of 100 archives takes us to 3/4 of the limit already as-is. The "caching" in `download_artifacts` _would_ likely save most of the downloads, but the workflow runs on GH's cloud runners, so it's a new machine every time, with nothing cached.

Do feel free to dismiss my nack if you agree with Radek - I've stated my case, that is enough.

Reply:
No, 37 pages should only count as 37 requests :) We only download the artifacts we need for the week, and this only runs once a week. Currently it's processing 3 artifacts for each day so it should be a max of 58 a week.